File: DFS/200.Number-of-Islands/200.Number-of-Islands.ts
function numIslands(grid: string[][]): number {
  const m = grid.length;
  const n = grid[0].length;
  let ret = 0;
  for (let i = 0; i < m; i++) {
    for (let j = 0; j < n; j++) {
      if (grid[i][j] !== '1') {
        continue;
      } else {
        dfs(i, j);
        ret += 1;
      }
    }
  }
  return ret;
  function dfs(m: number, n: number) {
    if (m < 0 || m >= grid.length || n < 0 || n >= grid[0].length) {
      return;
    }

    if (grid[m][n] === '1') {
      grid[m][n] = '2';
    } else {
      return;
    }
    dfs(m - 1, n);
    dfs(m + 1, n);
    dfs(m, n - 1);
    dfs(m, n + 1);
  }
}

File: DFS/200.Number-of-Islands/200.Number-of-Islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        result = 0
        m = len(grid)
        n = len(grid[0])
        
        def dfs(row, column):
            if row < 0 or row >= m  or column < 0 or column >= n:
                return 

            if grid[row][column] == '0':
                return
            
            grid[row][column] = '0'
            directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]
            for [x, y] in directions:
                dfs(row + x, column + y)

        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1':
                    result += 1
                    dfs(i, j)
        
        return result

                    
